kore/src/SMT/AST.hs

```diff
@@ -0,0 +1,220 @@
+                 ( Alternative (..) )
+import           Control.DeepSeq
+                 ( NFData )
+import           Control.Monad.Fail
+                 ( MonadFail )
+import           Data.Char
+                 ( isSpace )
+import           Data.Maybe
+                 ( fromMaybe )
+import           Data.Text
+                 ( Text )
+import           Data.Text.Internal.Builder
+                 ( Builder )
+import qualified Data.Text.Internal.Builder as Text.Builder
+import qualified Data.Text.IO as Text
+import qualified Data.Text.Lazy as Text.Lazy
+import           Data.Void
+                 ( Void )
+import           GHC.Generics
+                 ( Generic )
+import           System.IO
+                 ( Handle, hPutChar )
+import           Text.Megaparsec
+                 ( Parsec )
+import qualified Text.Megaparsec as Parser
+import qualified Text.Megaparsec.Char as Parser
+import qualified Text.Megaparsec.Char.Lexer as Lexer
+
+-- | S-expressions, the basic format for SMT-LIB 2.
+data SExpr
+    = Atom !Text
+    | List ![SExpr]
+    deriving (Generic, Eq, Ord, Show)
+
+{--| An argument to a data type constructor.
+
+The name can be used as a getter in smt solvers. (Note: this is currently not
+working due to a bug in z3 data type declaration,
+see https://github.com/Z3Prover/z3/issues/2217 , also see the similar comment
+in declareDatatypes in SimpleSMT.hs)
+--}
+data ConstructorArgument sort name =
+    ConstructorArgument
+        { name :: !name
+        , argType :: !sort
+        }
+    deriving (Eq, Ord, Show)
+
+{--| A data type constructor.
+--}
+data Constructor sort symbol name =
+    Constructor
+        { name :: !symbol
+        , arguments :: ![ConstructorArgument sort name]
+        }
+    deriving (Eq, Ord, Show)
+
+{--| A constructor-based data type declaration.
+
+If the list of constructors is empty, the data type is empty.
+--}
+data DataTypeDeclaration sort symbol name =
+    DataTypeDeclaration
+        { name :: !name
+        , typeArguments :: ![name]
+        , constructors :: ![Constructor sort symbol name]
+        }
+    deriving (Eq, Ord, Show)
+
+-- | A non-constructor-based data type declaration.
+data SortDeclaration name =
+    SortDeclaration
+        { name :: !name
+        , arity :: Int
+        }
+    deriving (Eq, Ord, Show)
+
+-- | A function declaration.
+data FunctionDeclaration sort name =
+    FunctionDeclaration
+        { name :: !name
+        , inputSorts :: ![sort]
+        , resultSort :: !sort
+        }
+    deriving (Eq, Ord, Show)
+
+-- | Extracts the name from a sexpression denoting a named object.
+nameFromSExpr :: SExpr -> Text
+nameFromSExpr (Atom name) = name
+nameFromSExpr (List (Atom name : _)) = name
+nameFromSExpr e =
+    (error . unlines)
+        [ "Cannot extract name from s-expression."
+        , "expression=" ++ showSExpr e
+        ]
+
+-- | Instantiate Constructor with the types needed by SimpleSMT.
+type SmtConstructor = Constructor SExpr Text Text
```

Comment:
Are these types ever used any other way?

Reply:
Yes. See the `UnresolvedConstructor` type in the other AST file. It is used in, say, Representation/Resolve.hs (which uses both UnresolvedConstructor and SmtConstructor and, which, most likely, uses all Smt* and Unresolved* types).